### PR TITLE
Add Hessian to FE_NedelecSZ

### PR DESCRIPTION
**Pull request by kinnewig**
Adds the Hessian to FE_NedelecSZ, as defined in the dissertation of S. Zaglmayr, "High Order Finite Element Methods for Electromagnetic Field Computation".

### CODE REVIEW

**Comment by Sebastian-47** [comment]
Thanks for the suggestion with the test! I just adapted the test fe_nedelec_hessian_divergence_theorem.cc to FE_NedelecSZ.

The implementation of the Hessian passed the test for the 2D case, but there still seems to be a problem in the 3D case. So I will take a closer look again at the Hessian in 3D and try to figure out the problem and update this pull request once I have fixed the problem.

**Comment by bangerth** [comment]
If you have a test that is based on integral identities such as
```
  \int_K div Phi = \int_{dK} Phi.n
```
then you can run this for an arbitrarily generated finite element function `Phi`, or you can run this for `Phi` being one shape function after the other. If one shape function is wrongly implemented and all of the others are correct, it will be relatively easy to figure out which one it is. Maybe you can apply this principle to your search if you used a similar relationship as above that involves the Hessian.

**Comment by Sebastian-47** [comment]
I finally found some time to work on this problem again.

It was very nasty to find the error in the fill_face_values function. I ended up recalculating the Hessian value again and re-implementing it. I also double and triple checked everything related to the calculation of the Hessian, so I'm pretty sure this should be correct.

**Comment by agrayver** [comment]
@Sebastian-47 fantastic! You will also need to push the test output file (it will be used a reference).

**Comment by Sebastian-47** [comment]
I just uploaded the results of the test fe_nedelec_sz_hessian_divergence_theorem.
Furthermore I resolved the merge conflict with #13150

**Comment by bangerth** [comment]
Out of curiosity, what was the bug?

**Comment by Sebastian-47** [comment]
> Out of curiosity, what was the bug?

Actually, I'm not quite sure where the problem was originally.

There was a minor mistake in function get_data, around line 1250, but that was easy to find and easy to fix as it was only a typo.

The problem was in the 3D-case of the function fill_fe_face_values. On the first try, I apparently made a mistake in the calculation of the hesssian for the face based functions. Therefore, I recalculated the hessian for the face based function again, and re-implemented it. 
I only briefly compared the two calculations, but I was not able to spot the mistake in the first one. After that, I just stopped trying to find the mistake on the first computation.

File: source/fe/fe_nedelec_sz.cc [hunk 1]
```diff
@@ -424,12 +426,12 @@ FE_NedelecSZ<dim, spacedim>::get_data(
...
                   // We only need poly values and 1st derivative for
                   // update_values, but need the 2nd derivative too for
-                  // update_gradients.
-                  //
-                  // Note that this will need to be updated if we're supporting
-                  // update_hessians.
-                  const unsigned int poly_length(
-                    (flags & update_gradients) ? 3 : 2);
+                  // update_gradients. For update_hessians we also need 3rd
+                  // derivative.
```
[the remark this answers is not in the document]
Comment (bangerth) [reply source/fe/fe_nedelec_sz.cc:431]: No, not fixed yet :-)

**Review by bangerth — COMMENTED** [review]
OK if you fix the one issue that's still open and merge the commits into either just one commit, or one commit for the code changes, and one commit for the test + test output.

Also: Would you mind adding an entry to `doc/news/changes/minor` in the same style as used there? This would also make sure that you get the credit you deserve!

**Comment by Sebastian-47** [comment]
I fixed that one open issue and merged all commits into just one commit.
Also, I added an entry to `doc/news/changes/minor`

**Review by bangerth — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]

**Comment by bangerth** [comment]
Nice work, @kinnewig !

**Pull request merged**